cost_reports/app.py

```diff
@@ -0,0 +1,186 @@
+                when = datetime.datetime(year=int(found.group(1)),
+                                         month=int(found.group(2)),
+                                         day=1)
+                self._year_month(when)
+            else:
+                LOG.error("Year/month could not be found in S3 path")
+
+    def _year_month(self, when=None):
+        """Return a string in the formart `year=XXXX/month=XX`
+        representing the given `datetime` (or current month if None)."""
+
+        now = datetime.datetime.now()
+        if when is None:
+            when = now
+
+        # First create the string used in the S3 path
+        self.year_month = f"year={when.year}/month={when.month}"
+
+        # Then create the string used in the report title
+        # format syntax: https://docs.python.org/3.9/library/time.html?highlight=time.strftime#time.strftime
+        self.title_date = when.strftime('%B %Y')
+        if when.month == now.month:
+            LOG.info("Processing month-to-date report")
+            self.title_date += ' (Month-to-Date)'
+        else:
+            LOG.info("Processing completed report")
+            self.title_date += ' (Complete)'
+
+
+    def _build_cur_key(self, s3_file):
+        """Build the path to the parquet file for last month"""
+
+        s3_path = f"{CUR_S3_PREFIX}/{REPORT_NAME}/{REPORT_NAME}/{self.year_month}"
+
+        s3_key = f"{s3_path}/{s3_file}"
+        LOG.info(f"Built S3 key: {s3_key}")
+        return s3_key
+
+    def _download_cur_parquet(self):
+        """Download the cost-and-usage report for the current month"""
+
+        s3_resp = App.s3.get_object(Bucket=CUR_S3_BUCKET, Key=self.cur_key)
+        self.cur_parquet = io.BytesIO(s3_resp['Body'].read())
+
+    def _download_account_csv(self):
+        """Download a TSV mapping accounts to cost centers"""
+
+        s3_resp = App.s3.get_object(Bucket=CUR_S3_BUCKET, Key=ACCOUNT_TSV_S3_KEY)
+        self.account_csv = io.BytesIO(s3_resp['Body'].read())
+
+    def _process_report(self):
+        """Group costs by program cost center"""
+
+        if self.cur_parquet is None:
+            raise Exception("No cost and usage data loaded")
+
+        if self.account_csv is None:
+            raise Exception("No account cost center data loaded")
+
+        self.summary, self.data = finops.main(self.cur_parquet, self.account_csv, self.title_date)
+        LOG.info("Summary generated")
+
+    def _upload_results(self):
+        """Upload the per-cost-center summary back to the S3 bucket"""
+
+        # Don't write an empty file
+        if self.summary is None:
+            LOG.error("No summary generated")
+            raise Exception("No summary generated")
+        if self.data is None:
+            LOG.error("No data generated")
+            raise Exception("No data generated")
+
+        summary_key = self._build_cur_key("summary.html")
+        App.s3.put_object(Bucket=CUR_S3_BUCKET, Key=summary_key, Body=self.summary)
```

Comment:
This has the effect of updating the 'summary' file if it has already been created, correct?

Reply:
yes, this will overwrite an existing file if it exists for the given year/month